routers/web/repo/repo.go

```diff
@@ -455,9 +457,30 @@ func download(ctx *context.Context, archiveName string, archiver *repo_model.Rep
 		return
 	}
 	defer fr.Close()
+	addCommitObjectResponseHeader(ctx, archiver)
 	ctx.ServeStream(fr, downloadName)
 }
 
+// Add X-Commit-Object response header if it was requested.
+// The commit object is needed to verify a downloaded archive by commit hash.
+// We use base64 to ensure a lossless encoding of binary data.
+// test: gitea/integrations/repo_download_test.go
+func addCommitObjectResponseHeader(ctx *context.Context, archiver *repo_model.RepoArchiver) {
+	if ctx.Req.Header.Get("X-Commit-Object") != "1" {
```

Comment:
Using `X-Commit-Object` for different purpose in request and response is not a good idea.

HTTP has `Content-Length` in request and response, it has the same meaning.

But now use `X-Commit-Object` for different meaning. I think we can make it like this:

request:
``` 
X-Request-Extra-Info: commit-id
```

response:
```
X-Extra-Commit-ID: .....
```

If we want to add more options, we can extend: `X-Request-Extra-Info: commit-message, commit-author` ...

Reply:
> is not a good idea

why not?

> If we want to add more options

then we can add more headers

keep it simple : )

Reply:
> why not?

As explained, the `X-Commit-Object` has different meanings, will mislead users and developers.

> keep it simple : )

I don't think `X-Request-Extra-Info` +  `X-Extra-Commit-ID` is more complex than `X-Commit-Object` + `X-Commit-Object`.

Reply:
thanks for your opinion, but youre just wasting my time. bye : )

Reply:
unresolve

its interesting that [bad taste](https://neilonsoftware.com/difficult-people-on-software-projects/developers/the-incompetent/) and [hostage-taking](https://neilonsoftware.com/difficult-people-on-software-projects/developers/the-hostage-taker/) so often appear together.
probably both are facets of the same personality type (conservative)

in the real world, i would kill the tyrant, and install someone who is "less bad".
in the digital world? lets just pretend this never happened ... "muh ideals!"

Reply:
Are you talking about yourself?

For me, I feel fine. I submitted some PRs, approved many PRs and helped some contributors to improve their PRs, like `#17152`, `#17900`, `#17929`, etc.

If you don't like my suggestions, that's your choice. I won't waste time on non-sense communications.

We can keep this review open and let everyone read it.